### PR TITLE
New Optuna Finetuning Script

### PR DESCRIPTION
**Pull request by ivanleomk**
Fixed up a new Optuna finetuning script and comitted some of the initial results. It seems like in general

- `tanh` is a better activation function
- larger output dims give better results
- linear layers allow us to scale AUC with dataset size (Eg. 128,000 datapoints with 8 epochs) has the best results 

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit cd95d4663ebf40f33823a16deb62dc68901fc6eb.  | 
|--------|--------|

### Summary:
This PR enhances the Optuna fine-tuning script for the Quora embeddings application, expands the list of tunable parameters, introduces minimum and maximum learning rates and epochs, updates the objective function, changes the results file format to CSV, and improves exception handling.

**Key points**:
- Introduced a new Optuna fine-tuning script for the Quora embeddings application.
- Expanded the list of tunable parameters, including models, dense layer dimensions, activation functions, schedulers, dataset sizes, warmup steps, and batch sizes.
- Introduced minimum and maximum learning rates, epochs, and an option to freeze the embedding model.
- Updated the objective function for fine-tuning.
- Changed the results file format from JSON to CSV.
- Updated the main function to handle exceptions and write the results to the CSV file.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: applications/finetune-quora-embeddings/optimize.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+import pandas as pd
+import math
+from typing import List
+
+# GPU Configuration
+gpu_config = gpu.A100()
+
+# Finetuning Configuration ( Arrays are configurable parameters )
+MODELS = [
+    # "BAAI/bge-base-en-v1.5",
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py]: make this a suite of models

File: applications/finetune-quora-embeddings/optimize.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    "WhereIsAI/UAE-Large-V1"
+]
+DENSE_LAYER_DIMS = [2048]
+ACTIVATION_FUNCTIONS = ["Tanh"]  # This is a torch func, we use getattr to read it
+SCHEDULER = ["warmupconstant"]
+DATASET_SIZE = [128000]
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py]: we should make this the same suite as before in the non finetuning example, 1k,2k etc,...

File: applications/finetune-quora-embeddings/optimize.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+WARMUP_STEPS = [500]
+BATCH_SIZE = 64
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py]: this hsould be a parameter

Comment (jxnl) [reply applications/finetune-quora-embeddings/optimize.py]: 32, 64, 128, etc

also no learning rate?

File: applications/finetune-quora-embeddings/optimize.json [hunk 4]
```diff
@@ -0,0 +1,372 @@
+
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.json]: can you make a csv of this

File: applications/finetune-quora-embeddings/optimize.py (hunk 1 above)
Comment (ivanleomk) [reply applications/finetune-quora-embeddings/optimize.py]: Added a suite of 5 models

File: applications/finetune-quora-embeddings/optimize.py (hunk 2 above)
Comment (ivanleomk) [reply applications/finetune-quora-embeddings/optimize.py]: set it to be [1000,2000,4000,8000, 16000,32000,64000,128000]

File: applications/finetune-quora-embeddings/optimize.py (hunk 3 above)
Comment (ivanleomk) [reply applications/finetune-quora-embeddings/optimize.py]: I got an OOM error when working with 128 previously. I think 96 might be an option though.

File: applications/finetune-quora-embeddings/optimize.json (hunk 4 above)
Comment (ivanleomk) [reply applications/finetune-quora-embeddings/optimize.json]: Converted to csv file of `prev_results.json`

File: applications/finetune-quora-embeddings/optimize.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+]
+DATASET_SIZE = [1000, 2000, 4000, 8000, 16000, 32000, 64000, 128000]
+WARMUP_STEPS = [500, 1000, 1500, 2000]
+BATCH_SIZE = [32, 64, 96]
+MODEL_SAVE_PATH = "/output"
+OPTIMIZATION_METRIC_FUNC = roc_auc_score
+MIN_LEARNING_RATE = 1e-5
+MAX_LEARNING_RATE = 1e-3
+MIN_EPOCHS = 6
+MAX_EPOCHS = 15
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py:41]: we should try to set to 10 epochs.
unless we can figure out early stopping

File: applications/finetune-quora-embeddings/optimize.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    "warmupconstant",
+    "warmuplinear",
+    "warmupcosine",
+    "warmupcosinewithhardrestarts",
+]
+DATASET_SIZE = [1000, 2000, 4000, 8000, 16000, 32000, 64000, 128000]
+WARMUP_STEPS = [500, 1000, 1500, 2000]
+BATCH_SIZE = [32, 64, 96]
+MODEL_SAVE_PATH = "/output"
+OPTIMIZATION_METRIC_FUNC = roc_auc_score
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py:37]: change this to accuracy

File: applications/finetune-quora-embeddings/optimize.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+SCHEDULER = [
+    "constantlr",
+    "warmupconstant",
+    "warmuplinear",
+    "warmupcosine",
+    "warmupcosinewithhardrestarts",
+]
+DATASET_SIZE = [1000, 2000, 4000, 8000, 16000, 32000, 64000, 128000]
+WARMUP_STEPS = [500, 1000, 1500, 2000]
+BATCH_SIZE = [32, 64, 96]
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py:35]: suggest_int

File: applications/finetune-quora-embeddings/optimize.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+# Finetuning Configuration ( Arrays are configurable parameters )
+MODELS = [
+    "BAAI/bge-base-en-v1.5",
+    "WhereIsAI/UAE-Large-V1",
+    "thenlper/gte-small",
+    "intfloat/e5-small-v2",
+    "sentence-transformers/all-MiniLM-L12-v2",
+]
+DENSE_LAYER_DIMS = [256, 512, 1024, 2048]
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py:21]: suggest_int

File: applications/finetune-quora-embeddings/optimize.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        order_outputs=False,
+        return_exceptions=True,
+    ):
+        if isinstance(resp, Exception):
+            print(f"Encountered Exception of {resp}")
+            continue
+
+        results.extend(resp)
+
+    # Check if the file exists
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/optimize.py:248]: seperate saving data code, and just use pandas, dont need to use csv.DictWriter. 

pd.Dataframes(**dict) (or records)

**Review by jxnl — APPROVED** [review]
add more directory structure in future prs
